### PR TITLE
add new index for Packages

### PR DESCRIPTION
**Pull request by bxf12315**
issues: https://github.com/trustification/trustification/issues/786 and https://github.com/trustification/trustification/issues/784

### CODE REVIEW

File: bombastic/index/src/lib.rs [hunk 1]
```diff
@@ -1,3 +1,6 @@
+pub mod packages;
+pub mod sbom;
```
Comment (lulf) [inline bombastic/index/src/lib.rs:2]: I think the rest of this file should be moved to sbom.rs and removed from lib.rs

File: bombastic/index/src/packages/mod.rs [hunk 2]
```diff
@@ -0,0 +1,564 @@
...
+                        document.add_text(fields.license, s.to_string());
+                    }
+                    LicenseIdentifier::SpdxId(_) => (),
+                },
+                LicenseChoice::Expression(_) => (),
+            });
+        }
+
+        document.add_text(fields.classifier, component.component_type.to_string());
+        documents.push((package_id, document));
```
Comment (lulf) [inline bombastic/index/src/packages/mod.rs:241]: This should only add the document if package_id is non-empty too I think.

**Review by lulf — CHANGES_REQUESTED** [review]
Great work 🚀 , I left a few comments.

File: bombastic/index/src/sbom/mod.rs [hunk 3]
```diff
@@ -0,0 +1,924 @@
+// use core::str::FromStr;
```
Comment (ctron) [inline bombastic/index/src/sbom/mod.rs:1]: Is this file still needed?

File: bombastic/indexer/src/lib.rs [hunk 4]
```diff
@@ -58,6 +59,12 @@ impl Run {
                     let index = block_in_place(|| {
                         IndexStore::new(&self.storage, &self.index, sbom_index, context.metrics.registry())
                     })?;
+
+                    let package_index: Box<dyn WriteIndex<Document = (SBOM, String)>> =
+                        Box::new(packages::Index::new());
+                    let packageinfo_index = block_in_place(|| {
```
Comment (ctron) [inline bombastic/indexer/src/lib.rs:65]: I think it would be good to align the naming a bit. Maybe go with `*_index` and `*_store`?

File: index/src/lib.rs [hunk 5]
```diff
@@ -330,15 +330,19 @@ impl IndexWriter {
...
-                self.delete_document(index, id);
-                self.writer.add_document(doc).map_err(|e| {
+                let docs = index.index_doc(id, &doc).map_err(|e| {
                     self.metrics.failed_total.inc();
                     e
                 })?;
+                for (i, doc) in docs {
+                    self.delete_document(index, &i);
+                    self.writer.add_document(doc).map_err(|e| {
+                        let e = e.clone();
```
Comment (ctron) [inline index/src/lib.rs:340]: Is this clone required? or could the function simply return `e`?

File: v11y/index/src/lib.rs [hunk 6]
```diff
@@ -347,10 +349,10 @@ impl trustification_index::WriteIndex for Index {
         "cve"
     }
 
-    fn index_doc(&self, _id: &str, doc: &Cve) -> Result<Document, SearchError> {
+    fn index_doc(&self, _id: &str, doc: &Cve) -> Result<Vec<(String, Document)>, SearchError> {
         match doc {
-            Cve::Published(cve) => self.index_published_cve(cve),
-            Cve::Rejected(cve) => self.index_rejected_cve(cve),
+            Cve::Published(cve) => self.index_published_cve(cve, _id),
```
Comment (ctron) [inline v11y/index/src/lib.rs:354]: If `_id` is actually being used, it should be `id`.

**Review by ctron — COMMENTED** [review]
Great work indeed! Some minor comments from me.

File: bombastic/index/src/packages/mod.rs (hunk 2 above)
Comment (bxf12315) [reply bombastic/index/src/packages/mod.rs:241]: the id use in here https://github.com/trustification/trustification/blob/5f9b67b6bfe52b54b1f4f7afa2b1e0ee9110288d/index/src/lib.rs#L338. index will delete old doc before add new doc.

File: index/src/lib.rs (hunk 5 above)
Comment (bxf12315) [reply index/src/lib.rs:340]: if do not use e.clone(), the code will have the error "Use of moved value ".

Comment (ctron) [reply index/src/lib.rs:340]: Hm, I don't see where it could be moved. Maybe some old code that was there before?

Comment (bxf12315) [reply index/src/lib.rs]: it's intellij's mentioned error, after remove it, it's work. I will remove it.

File: bombastic/index/src/packages/mod.rs (hunk 2 above)
Comment (lulf) [reply bombastic/index/src/packages/mod.rs:241]: Right, but if it didn't find a purl within the component above, the package_id would have the value `""`, which would be invalid.

**Review by ctron — COMMENTED** [review]
LGTM … I will leave the +1 for @lulf

**Comment by bxf12315** [comment]
> Looks like the integration tests fail due to "duplicate metrics"

https://github.com/trustification/trustification/blob/6cdb80e9c38bedcfb586c615b65499d157776fed/bombastic/indexer/src/lib.rs#L59
https://github.com/trustification/trustification/blob/6cdb80e9c38bedcfb586c615b65499d157776fed/bombastic/indexer/src/lib.rs#L65
Added metrics twice in here, if we remove one of them, it will work. I think we needs to add different metrics here.
How to let indexer can support multiple metrics?

**Comment by bxf12315** [comment]
> > Looks like the integration tests fail due to "duplicate metrics"
> 
> https://github.com/trustification/trustification/blob/6cdb80e9c38bedcfb586c615b65499d157776fed/bombastic/indexer/src/lib.rs#L59
> 
> 
> https://github.com/trustification/trustification/blob/6cdb80e9c38bedcfb586c615b65499d157776fed/bombastic/indexer/src/lib.rs#L65
> 
> 
> Added metrics twice in here, if we remove one of them, it will work. I think we needs to add different metrics here.
> How to let indexer can support multiple metrics?

I add new metrics, but it's not a good idea.

**Comment by ctron** [comment]
> > Looks like the integration tests fail due to "duplicate metrics"
> 
> https://github.com/trustification/trustification/blob/6cdb80e9c38bedcfb586c615b65499d157776fed/bombastic/indexer/src/lib.rs#L59
> 
> https://github.com/trustification/trustification/blob/6cdb80e9c38bedcfb586c615b65499d157776fed/bombastic/indexer/src/lib.rs#L65
> 
> Added metrics twice in here, if we remove one of them, it will work. I think we needs to add different metrics here.
> How to let indexer can support multiple metrics?

I think it makes sense to scope them with a prefix maybe? @lulf what do you think?

File: bombastic/indexer/src/lib.rs [hunk 7]
```diff
@@ -53,11 +55,16 @@ impl Run {
...
-                        Box::new(bombastic_index::Index::new());
+                    let sbom_index: Box<dyn WriteIndex<Document = (SBOM, String)>> = Box::new(sbom::Index::new());
                     let index = block_in_place(|| {
                         IndexStore::new(&self.storage, &self.index, sbom_index, context.metrics.registry())
                     })?;
+
+                    let package_index: Box<dyn WriteIndex<Document = (SBOM, String)>> =
+                        Box::new(packages::Index::new());
+                    let package_store = block_in_place(|| {
+                        IndexStore::new(&self.storage, &self.index, package_index, &Registry::new())
```
Comment (ctron) [inline bombastic/indexer/src/lib.rs:66]: We need to fix this, as those metrics won't be exposed.

**Review by ctron — CHANGES_REQUESTED** [review]

File: bombastic/indexer/src/lib.rs (hunk 7 above)
Comment (bxf12315) [reply bombastic/indexer/src/lib.rs:66]: like this one https://github.com/tikv/rust-prometheus/blob/master/examples/example_custom_registry.rs#L24?

**Comment by lulf** [comment]
For the metrics I think this would make sense:

* In the register, pass the value from calling WriteIndex::name()
* Add the name as a dimension to the metric

File: bombastic/indexer/src/lib.rs (hunk 7 above)
Comment (lulf) [reply bombastic/indexer/src/lib.rs:66]: It is better to prefix the metrics in register() taking the name as a parameter. So that in Metrics::register() the registration becomes:


```
  fn register(registry: &Registry, prefix: &str) -> Metrics {
   let indexed_total = register_int_counter_with_registry!(
            opts!("{prefix}_index_indexed_total", "Total number of indexing operations"),
            registry
        )?;


```

**Comment by bxf12315** [comment]
> For the metrics I think this would make sense:
> 
> * In the register, pass the value from calling WriteIndex::name()
> * Add the name as a prefix when registering the metric

I updated the PR. How about this one?

File: bombastic/indexer/src/lib.rs (hunk 7 above)
Comment (bxf12315) [reply bombastic/indexer/src/lib.rs]: after create register with prefix, the prefix will be in here.  https://github.com/tikv/rust-prometheus/blob/master/examples/example_custom_registry.rs#L24

**Review by lulf — COMMENTED** [review]
The problem is that custom_registry() will create a new registry that is not connected to the HTTP endpoint, so although it will allow metrics to be registered, they will not be surfaced to the HTTP endpoint, which is done in the infra.rs (line 101):

```
async fn metrics(metrics: web::Data<Arc<Metrics>>) -> HttpResponse {
    let encoder = TextEncoder::new();
    let metric_families = metrics.registry().gather();
    match encoder.encode_to_string(&metric_families) {
        Ok(data) => HttpResponse::Ok().content_type("text/plain").body(data),
        Err(e) => HttpResponse::InternalServerError().body(format!("Error retrieving metrics: {:?}", e)),
    }
}
```

So either the registries must be added to a Vec<Registry> in infra.rs or the same registry must be used.

File: bombastic/indexer/src/lib.rs (hunk 7 above)
Comment (ctron) [reply bombastic/indexer/src/lib.rs]: I believe that creating a new registry would also mean that an additional one needs to be exposed. So re-using the same instance makes more sense to me.

Comment (ctron) [reply bombastic/indexer/src/lib.rs]: Sorry, Ulf already wrote that "below".

**Review by lulf — APPROVED** [review]
Looks good, great work!

**Review by ctron — APPROVED** [review]
